Implementation of delta-sync support on server-side.

ahmedammar opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
This commit adds server-side support for delta-sync, this extends the core files app.

The basic approach is to store zsync metadata files in a folder called `files_zsync/` which mirrors the structure and layout of the `files/` folder but appends `.zsync` to the metadata files. These files can be requested by the client via a new route `api/v1/zsync/$path`.

Filesystem hooks are used to mirror any `move/copy/delete` operation on the base file onto the metadata file.

The upload path is implemented by modifying the `ChunkingPlugin`, the chunk files are now assumed to be named as the offsets into the original file. The implemenation copies adds a new class `AssemblyStreamZsync` which extends `AssemblyStrem` with support to fill in the missing data in between chunk offsets from a `backingFile`.

A new `zsync` capability is added to the files app, which can be checked by the client to know if delta-sync is supported or not.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
owncloud/core#16162.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Mostly tested by using the owncloud command line client and macOS application. With some randomly generated files using `dd` and various modifications to those files:
```
dd if=/dev/urandom of=1g.rand.img bs=$[1024*1024] count=1024

cp 1g.rand.img 1g.rand.add300m.img
dd if=/dev/urandom of=1g.rand.add300m.img bs=$[1024*1024] count=300 oseek=1024

cp 1g.rand.img 1g.rand.mod200m.img
dd if=/dev/urandom of=1g.rand.mod200m.img bs=$[1024*1024] count=100 oseek=123 conv=notrunc 
dd if=/dev/urandom of=1g.rand.mod200m.img bs=$[1024*1024] count=100 oseek=532 conv=notrunc

cp 1g.rand.img 1g.rand.mov200m.img
dd if=1g.rand.img of=1g.rand.mov200m.img bs=$[1024*1024] count=200 iseek=118 oseek=418 conv=notrunc

cp 1g.rand.img 1g.rand.del200m.img
dd if=1g.rand.img of=1g.rand.del200m.img bs=$[1024*1024] count=622
dd if=1g.rand.img of=1g.rand.del200m.img bs=$[1024*1024] iseek=822 oseek=622
```

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/core?pullRequest=29378) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/core?pullRequest=29378) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/core?pullRequest=29378) it.</sub>

ahmedammar committed:
Implementation of delta-sync support on server-side.

This commit adds server-side support for delta-sync, this extends the
core files app.

The basic approach is to store zsync metadata files in a folder called
`files_zsync/` which mirrors the structure and layout of the `files/`
folder but appends `.zsync` to the metadata files. These files can be
requested by the client via a new route `api/v1/zsync/$path`.

Filesystem hooks are used to mirror any `move/copy/delete` operations on
the base file onto the metadata file.

The upload path is implemented by modifying the `ChunkingPlugin`, the
chunk files are now assumed to be named as the offsets into the original
file. The implemenation copies adds a new class `AssemblyStreamZsync`
which extends `AssemblyStrem` with support to fill in the missing data
in between chunk offsets from a `backingFile`.

A new `zsync` capability is added to the files app, which can be checked
by the client to know if delta-sync is supported or not.

This commit closes owncloud/core#1612.

jvillafanez commented:
Question to everyone involved: as far as I know, this is another chunking algorithm (or at least it involves modifying it); why isn't this in another app that can be enabled / disabled and why there isn't a switch to choose the algorithm?

I understand that there will be changes in core to support this, but those changes should be oriented to support more chunking algorithms. The algorithm itself should be moved to an application.

(I haven't been in the discussions so there might be important things that I missed)

ahmedammar commented:
@jvillafanez well I'm here trying to start a discussion :) 

I assumed that delta-sync should be part of the core server implementation and not a separate app. If that isn't the case, then it can be modified. 

The changes to the chunking algorithm are very minimal and don't break any previous client code.

jvillafanez commented:
Let's suppose we merge this PR now. What "business value" add this feature to the product and who can benefit from it now?

From my point of view, only the desktop client can benefit from this PR, which I guess it will help for faster syncing, but mobile clients nor the web UI will gain anything. In fact they might perform slower. Note that we also need to take into account external storages and where those ".zsync" files will be created.

This is why I think it's better to move it to an app so the admin can choose what he want to do.

ahmedammar commented:
Well I don't agree to be honest, only desktop will potentially benefit instantly, but nothing is in the way of a mobile implementation too.

Why would anything perform slower because of this PR? I fail to understand that at all.

Nothing in the PR affects existing code paths or breaks any legacy code. Without a supported client nothing changes.

ahmedammar commented:
I also fail to see how an 'app' can modify the ChunkingPlugin or even override it.

jvillafanez commented:
My point is that mobiles don't have this support implemented now, and it will take time to implement it (if ever, due to priorities). Meanwhile, the client will just upload the files as they have been doing all this time.

> Why would anything perform slower because of this PR?

In the ChunkingPlugin file, `performMove` function you've wrapped the "old" code `$this->server->tree->move($path, $destination);` with your new code. This will increase the cost of the regular uploads without benefit except for the desktop client (yes, mobiles can implement whatever is needed to gain performance with these changes, but meanwhile it's like this). How much cost is added there? Without measuring, I can't say for sure. Maybe 1 ms, 100 ms, 1 sec... maybe I'm raising unneeded alarms (I haven't reviewed the code in depth so maybe it just adds a couple of ms and it's nothing to worry about).

> The final TODO (if you looked at the related issue) is related to disabling for external storage.

Disabling or switching? Chunking is still needed, which reinforces the idea of 2 different implementations.

> I also fail to see how an 'app' can modify the ChunkingPlugin or even override it. So it may as well be part of the existing files app.

Probably it isn't possible at the moment, but this support can be added somehow.

ahmedammar commented:
> In the ChunkingPlugin file, performMove function you've wrapped the "old" code $this->server->tree->move($path, $destination); with your new code

I've re-ordered the code so it's a bit clearer, it's basically a string modification and two if statements. For anyone not using zsync based chunking that should be a few ms maximum. I'd be surprised otherwise. the regex string replace could be replaced with something faster if that's *really* a performance issue.

https://github.com/ahmedammar/core/commit/c7b22e5c430d52af9c2e83c1c88375e29598dddc#diff-a3fb8f83598c3f5e1d514a34ca578b1eR86

> Disabling or switching? Chunking is still needed.

No one said anything about disabling Chunking at all ... Disabling zsync metadata upload and fetch for external storage is what I meant.

> Probably it isn't possible at the moment, but this support can be added somehow.

I personally find it hard to believe that multiple chunking algorithms/plugins is even something with enough developer need. The one **important** reason this modification of ChunkingPlugin needed to happen for zsync metadata is to get around the fact that there is no persistent file locking available in owncloud. If the .zsync is not copied in the same MOVE request as the base file then this leaves the door open for issues with multiple uploaders.

ahmedammar commented:
I closed this because I accidentally deleted my repo on GitHub 😞 

I opened a new PR#29404.

@jvillafanez Please continue the discussion there
